File: algorithms/sorting/quicksort/quicksort.py
def partition(array, left_index, right_index):
    pivot_index = right_index
    pivot = array[right_index]
    right_index -= 1

    while True:
        while array[left_index] < pivot:
            left_index += 1
        
        while array[right_index] > pivot:
            right_index -= 1
        
        if left_index >= right_index:
            break
        else:
            array[left_index], array[right_index] = array[right_index], array[left_index]
            left_index += 1
        
    array[left_index], array[pivot_index] = array[pivot_index], array[left_index]
    
    return left_index

def quicksort(array, left_index=0, right_index=None):
    if right_index is None:
        right_index = len(array) - 1
    
    if right_index - left_index <= 0:
        return

    pivot_index = partition(array, left_index, right_index)
    quicksort(array, left_index, pivot_index - 1)
    quicksort(array, pivot_index + 1, right_index)

    return array

File: algorithms/sorting/quicksort/quicksort.js
class QuickSort {
    constructor(array) {
        this.array = array;
    }


    partition(leftIndex, rightIndex) {
        let pivotIndex = rightIndex;
        let pivot = this.array[pivotIndex];
        rightIndex--;
        
        while(true) {

            while (this.array[leftIndex] < pivot) {
                leftIndex++;
            }

            while (this.array[rightIndex] > pivot) {
                rightIndex--;
            }

            if (leftIndex >= rightIndex) {
                break;
            }
            else {
                let tempLeft = this.array[leftIndex];
                this.array[leftIndex] = this.array[rightIndex];
                this.array[rightIndex] = tempLeft;
                leftIndex++;
            }
        }

        let tempLeft = this.array[leftIndex];
        this.array[leftIndex] = pivot;
        this.array[pivotIndex] = tempLeft;
        
        return leftIndex;
    }
    
    quicksort(leftIndex, rightIndex) {
        if (rightIndex - leftIndex <= 0) {
            return;
        }

        let pivotIndex = this.partition(leftIndex, rightIndex);
        this.quicksort(leftIndex, pivotIndex - 1);
        this.quicksort(pivotIndex + 1, rightIndex);
    }

    sort() {
        this.quicksort(0, this.array.length - 1);
        return this.array;
    }
}